salt/modules/virt.py

```diff
@@ -574,96 +592,39 @@ def init(name,
     '''
+
     hypervisor = __salt__['config.get']('libvirt:hypervisor', hypervisor)
     log.debug('Using hyperisor {0}'.format(hypervisor))
 
     nicp = _nic_profile(nic, hypervisor, **kwargs)
     log.debug('NIC profile is {0}'.format(nicp))
 
-    diskp = None
-    seedable = False
-    if image:  # with disk template image
-        log.debug('Image {0} will be used'.format(image))
-        # if image was used, assume only one disk, i.e. the
-        # 'default' disk profile
-        # TODO: make it possible to use disk profiles and use the
-        # template image as the system disk
-        diskp = _disk_profile('default', hypervisor, **kwargs)
-        log.debug('Disk profile is {0}'.format(diskp))
-
-        # When using a disk profile extract the sole dict key of the first
-        # array element as the filename for disk
+    diskp = _disk_profile(disk, hypervisor, **kwargs)
+
+    if image:
+        # Backward compatibility: if 'image' is specified in the VMs arguments
+        # instead of a disk arguments. In this case, 'image' will be assigned
+        # to the first disk for the VM.
         disk_name = next(six.iterkeys(diskp[0]))
-        disk_type = diskp[0][disk_name]['format']
-        disk_file_name = '{0}.{1}'.format(disk_name, disk_type)
-
-        if hypervisor in ['esxi', 'vmware']:
-            # TODO: we should be copying the image file onto the ESX host
-            raise SaltInvocationError(
-                'virt.init does not support image template template in '
-                'conjunction with esxi hypervisor'
-            )
-        elif hypervisor in ['qemu', 'kvm']:
-            img_dir = __salt__['config.option']('virt.images')
-            log.debug('Image directory from config option `virt.images` is {0}'
-                      .format(img_dir))
-            img_dest = os.path.join(
-                img_dir,
-                name,
-                disk_file_name
-            )
-            log.debug('Image destination will be {0}'.format(img_dest))
-            img_dir = os.path.dirname(img_dest)
-            log.debug('Image destination directory is  {0}'.format(img_dir))
-            sfn = __salt__['cp.cache_file'](image, saltenv)
-
-            try:
-                os.makedirs(img_dir)
-            except OSError:
-                pass
-
-            qcow2 = False
-            if salt.utils.which('qemu-img'):
-                res = __salt__['cmd.run']('qemu-img info {}'.format(sfn))
-                imageinfo = yaml.load(res)
-                qcow2 = imageinfo['file format'] == 'qcow2'
-
-            try:
-                if enable_qcow and qcow2:
-                    log.info('Cloning qcow2 image {} using copy on write'
-                              .format(sfn))
-                    __salt__['cmd.run'](
-                        'qemu-img create -f qcow2 -o backing_file={} {}'
-                        .format(sfn, img_dest).split())
+        if not diskp[0][disk_name].get('image', None):
```

Comment:
Default disk profile doesn't have 'image' attribute, so will always use 'image' from the module arguments.

But if a disk profile uses different images for different disks, there is no determined way to specify which images should be replaced with the image from argument:
disk_a:
    image: aaaa.qcow2
disk_b:
   image: bbbb.qcow2
disk_c:
   image: cccc.qcow2

If you think that the 'image' module argument should be for the first disk only, then it should be reflected at least in the docstring